sim/midas/src/main/cc/simif_token_hashers.cc

```diff
@@ -0,0 +1,222 @@
+#endif
+}
+
+/**
+ * Print debug info about the MMIO internals
+ */
+void simif_token_hashers_t::info() {
+  std::cout << "trigger0 " << trigger0 << "\n";
+  std::cout << "trigger1 " << trigger1 << "\n";
+  std::cout << "period0 " << period0 << "\n";
+  std::cout << "period1 " << period1 << "\n";
+
+  for (uint32_t i = 0; i < cnt; i++) {
+    std::cout << "i: " << i << "\n";
+    std::cout << "  bridge: " << bridge_names[i] << "\n";
+    std::cout << "  name: " << names[i] << "\n";
+    std::cout << "  direction: " << (outputs[i] ? "Output" : "Input") << "\n";
+    std::cout << "  queue_head: " << queue_heads[i] << "\n";
+    std::cout << "  queue_occupancy: " << queue_occupancies[i] << "\n";
+    std::cout << "  tokencount0: " << tokencounts0[i] << "\n";
+    std::cout << "  tokencount1: " << tokencounts1[i] << "\n";
+  }
+}
+
+/**
+ * Set the delay and period for the token hashers
+ *
+ * @param [in] delay The number of tokens before hashes are saved
+ * @param [in] period The number of hashes to skip between saving. 0 means save
+ * every hash
+ */
+void simif_token_hashers_t::set_params(const uint64_t delay,
+                                       const uint64_t period) {
+#ifndef TOKENHASHMASTER_0_PRESENT
+  std::cout << "simif_token_hashers_t::set_params() was called but Token Hashers are not enabled in this build\n";
+  return;
+#endif  
+  parent->write(trigger0, (delay & 0xffffffff));
+  parent->write(trigger1, ((delay >> 32) & 0xffffffff));
+
+  parent->write(period0, (period & 0xffffffff));
+  parent->write(period1, ((period >> 32) & 0xffffffff));
+}
+
+/**
+ * Readout all token hashes from the FPGA using MMIO.
+ * Returned value is a vector of vector of hashes. The index
+ * to the outer vector is the number of the bridge. The bridge oder / numbering
+ * is determined by scala at compiletime. Calling this twice in a row 
+ * will yield empty results the second time. This is because
+ * the FIFO's are drained and so occupancy will return 0
+ * @retval a vector of vector of hashes
+ */
+token_hasher_result_t simif_token_hashers_t::get() {
+  token_hasher_result_t ret;
+  for (uint32_t i = 0; i < cnt; i++) {
+    ret.push_back({});
+    const uint32_t occ = occupancy(i);
+    std::vector<uint32_t> &data = ret[i];
+    data.reserve(occ);
+    for (uint32_t j = 0; j < occ; j++) {
+      const uint32_t h = parent->read(queue_heads[i]);
+      data.push_back(h);
+    }
+  }
+
+  return ret;
+}
+
+/**
+ * Same as get() however multiple calls will alwaus return the same data
+ * Work is only done on the first call, subsequent calls return cached data
+ * from the first time.
+ */
+token_hasher_result_t simif_token_hashers_t::cached_get() {
+  if(cached_results.size() == 0) {
+    cached_results = get();
+  }
+
+  return cached_results;
+}
+
+/**
+ * Get a string of all the hashes. This calls cached_get() internally
+ * @retval a std::string with human readable output
+ */
+std::string simif_token_hashers_t::get_string() {
```

Comment:
Instead of returning a stream, parameterize this with a stream (`std::ostream &os`) and dump stuff to it. I see that you are dumping the resulting string to stdout, I imagine the contents of this could become quite large. We should avoid constructing the string altogether.